### PR TITLE
EBANX: Introducing EBANX's V2 gateway

### PR DESCRIPTION
**Pull request by leosteil**
## Summary:

EBANX: introducing EBANX's new version gateway. 
In EBANX side was created a new structure to work with all connectors, as Spreedly. Then, we created V2 gateway to work with this new approach.

### Remote Test:
Finished in 29.754991 seconds.
33 tests, 105 assertions, 0 failures, 0 errors, 0 pendings,0 omissions, 0 notifications 100% passed

### Unit Tests:
Finished in 0.04074 seconds.
18 tests, 99 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications 100% passed

### RuboCop:
760 files inspected, no offenses detected

### CODE REVIEW

**Comment by leosteil** [comment]
> Hello @leosteil thank you for creating a PR for this update. We've created a ticket to track the effort of reviewing it and will get to it soon. In the meantime could you answer a few questions
> 
> * What is the strategy for a merchant when deciding which version to use?
> * How should a merchant or integrator tell ActiveMerchant which version to use and is this backwards compatible?
> * What advantages does the V2 offer over V1?
> 
> Thanks!

Hi @aenand. thank you for your attention.

Answering your questions:

- What is the strategy for a merchant when deciding which version to use?

The decision of which version to use will be up to EBANX and will be transparent to the merchant. Initially, only a test account will use V2, until testing is complete and other merchants are directed to it.

- How should a merchant or integrator tell ActiveMerchant which version to use and is this backwards compatible?

There is no need for a merchant or integrator to tell ActiveMerchant. The decision will be managed be EBANX. V2 will be backwards compatible

- What advantages does the V2 offer over V1?

V2's architecture allows for faster implementation of new payment methods in the future, with reduced effort for both Spreedly and EBANX.

**Comment by aenand** [comment]
@leosteil ah gotcha, is there any concern on the latency added by having an additional API request that will reach out to Ebanx to get the gateway version?

**Comment by leosteil** [comment]
@aenand we made tests to get the average latency of requesting our api to get gateway version and it took about 400 ms

**Review by aenand — CHANGES_REQUESTED** [review]
The test coverage on this looks great! I have a couple questions to better understand the goal so I can provide some more detailed feedback:

- I know that EBANX wants to be able to make changes and experiment on their end to move quickly, what are the types of changes you want to be able to make? authentication or field level requests or data tracking or something else?
- Spreedly extends this EBANX class in our repository and this introduces a breaking change for our LPM integration at the moment since we are taking the ebanx_base_gateway and extending our lpm library on top of it. What are the deciding factors for determining v1 vs v2 (i.e. transaction type? payment method type? something else?)

File: lib/active_merchant/billing/gateways/ebanx.rb [hunk 1]
```diff
@@ -42,196 +20,128 @@ class EbanxGateway < Gateway
...
-        post[:operation] = 'request'
+        if @gateway_version === 'v1'
+          transcript.
+            gsub(/(integration_key\\?":\\?")(\w*)/, '\1[FILTERED]').
+            gsub(/(card_number\\?":\\?")(\d*)/, '\1[FILTERED]').
+            gsub(/(card_cvv\\?":\\?")(\d*)/, '\1[FILTERED]')
+        else
+          transcript.
+            gsub(/(number\\?":\\?")(\d*)/, '\1[FILTERED]').
+            gsub(/(verification_value\\?":\\?")(\d*)/, '\1[FILTERED]')
```
[the remark this answers is not in the document]
Comment (leosteil) [reply lib/active_merchant/billing/gateways/ebanx.rb:104]: In V2, integration key will be sent in Authorization Header, not in request body, as V1

File: lib/active_merchant/billing/gateways/ebanx/ebanx_v1.rb [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        if country = options[:country] || (options[:billing_address][:country] if options[:billing_address])
+          country.downcase
+        end
+      end
+
+      private
+
+      def add_integration_key(post)
+        post[:integration_key] = @integration_key
+      end
```
[the remark this answers is not in the document]
Comment (leosteil) [reply lib/active_merchant/billing/gateways/ebanx/ebanx_v1.rb:157]: As answered in gsub question, integration key in V2 will be sent in Authorization Header

**Comment by aenand** [comment]
@leosteil Spreedly extends this EBANX class in our repository and this introduces a breaking change for our LPM integration at the moment since we are taking the ebanx_base_gateway and extending our lpm library on top of it. What are the deciding factors for determining v1 vs v2 (i.e. transaction type? payment method type? something else?)

**Comment by leosteil** [comment]
- I know that EBANX wants to be able to make changes and experiment on their end to move quickly, what are the types of changes you want to be able to make? authentication or field level requests or data tracking or something else?
- Spreedly extends this EBANX class in our repository and this introduces a breaking change for our LPM integration at the moment since we are taking the ebanx_base_gateway and extending our lpm library on top of it. What are the deciding factors for determining v1 vs v2 (i.e. transaction type? payment method type? something else?)

1 - We want to be able to add new payment methods without the need of making changes in ebanx gateway at active_merchant. When Spreedly start working with Alternative Payment Methods (APMs) with ebanx gateway, We imagine that the necessary information about that will be sent for Us in the rawest way possible, then We can produce a payment with APMS in our side. The tracking will still the same as V2 and the authentication too, only changing the way of send the integration key.
2 - In this first moment, V2 will only run for a Demonstration Merchant that EBANX uses for testing. After we feel confident that we not introduced any bugs in V2 against V1 payments, We will enable V2 for all mechants. We will not add any merchat either change any business rule before talking with Spreedly Team

**Comment by leosteil** [comment]
Ok, got it. So, do you think is it not necessary the v2 structure to process LPMs? About a POC, a dont see problems about that.

**Comment by aenand** [comment]
After discussing with Ebanx we agreed this is not needed at the time

**Pull request closed without merging**